### PR TITLE
[Runtimes] Support loading nuclio functions from external code entry

### PR DESCRIPTION
**Pull request by sahare92**

### CODE REVIEW

File: mlrun/runtimes/function.py [hunk 1]
```diff
@@ -167,6 +168,114 @@ def add_trigger(self, name, spec):
...
+        git:
+        ("git://github.com/org/repo#my-branch",
+         handler="path/inside/repo#main:handler",
+         credentials={"GIT_PASSWORD": "my-access-token"})
+        s3:
+        ("s3://my-bucket/path/in/bucket/my-functions-archive",
+         handler="path/inside/functions/archive#main:Handler",
+         runtime="golang",
+         credentials={"AWS_ACCESS_KEY_ID": "some-id", "AWS_SECRET_ACCESS_KEY": "some-secret"})
+        """
```
Comment (yaronha) [inline mlrun/runtimes/function.py:197]: @sahare92 i dont think the doc string will render ok (e.g. code examples should be idented and after "::", u should check it

File: mlrun/runtimes/function.py [hunk 2]
```diff
@@ -167,6 +168,114 @@ def add_trigger(self, name, spec):
         self.spec.config[f"spec.triggers.{name}"] = spec
         return self
 
+    def from_remote_source(
+        self, source, handler="", runtime="", credentials=None,
+    ):
```
Comment (yaronha) [inline mlrun/runtimes/function.py:173]: use `secrets` for consistency w other calls

File: mlrun/runtimes/function.py [hunk 3]
```diff
@@ -167,6 +168,114 @@ def add_trigger(self, name, spec):
...
+                 secrets={"GIT_PASSWORD": "my-access-token"})
+            s3:
+                ("s3://my-bucket/path/in/bucket/my-functions-archive",
+                 handler="path/inside/functions/archive#main:Handler",
+                 runtime="golang",
+                 secrets={"AWS_ACCESS_KEY_ID": "some-id", "AWS_SECRET_ACCESS_KEY": "some-secret"})
+        """
+        code_entry_type = self._resolve_code_entry_type(source)
+        if code_entry_type == "":
+            raise ValueError("Couldn't resolve code entry type from source")
```
Comment (Hedingber) [inline mlrun/runtimes/function.py:200]: Use `mlrun.errors.MLRunInvalidArgumentError`

File: mlrun/runtimes/function.py [hunk 4]
```diff
@@ -167,6 +168,114 @@ def add_trigger(self, name, spec):
...
+        work_dir, handler = self._resolve_work_dir_and_handler(handler)
+        if work_dir != "":
+            code_entry_attributes["workDir"] = work_dir
+
+        if secrets is None:
+            secrets = {}
+
+        # set default runtime if not specified otherwise
+        if runtime == "":
+            runtime = "python:3.7"
```
Comment (Hedingber) [inline mlrun/runtimes/function.py:214]: Make configurable (add to `config.py` `default_nuclio_runtime`)

File: mlrun/runtimes/function.py [hunk 5]
```diff
@@ -167,6 +168,114 @@ def add_trigger(self, name, spec):
...
+
+        # archive
+        if code_entry_type == "archive":
+            if source.startswith("v3io"):
+                source = f"http{source[4:]}"
```
Comment (Hedingber) [inline mlrun/runtimes/function.py:219]:
```suggestion
                source = f"http{source[len("v3io"):]}"
```

File: mlrun/runtimes/function.py [hunk 6]
```diff
@@ -167,6 +168,114 @@ def add_trigger(self, name, spec):
...
+                code_entry_attributes["headers"] = {
+                    "headers": {"X-V3io-Session-Key": v3io_access_key}
+                }
+
+        # s3
+        if code_entry_type == "s3":
+            try:
+                bucket, item_key = parse_s3_bucket_and_key(source)
+            except Exception as exc:
+                raise RuntimeError(f"failed to parse s3 bucket and key. {exc}")
```
Comment (Hedingber) [inline mlrun/runtimes/function.py:234]: `mlrun.errors.MLRunInvalidArgumentError`

Comment (Hedingber) [reply mlrun/runtimes/function.py:234]: Actually, it looks like the only place you're using it, I don't see a reason not raising inside the function itself

**Review by Hedingber — CHANGES_REQUESTED** [review]

**Review by Hedingber — CHANGES_REQUESTED** [review]
@sahare92 PR looks good, 2 last things:
1. it seems that you have a conflict with upstream
2. @yaronha did some name changes in [his PR](https://github.com/mlrun/mlrun/pull/888) e.g. `with_source_archive` instead of `from_remote_source` - please change accordingly

**Review by Hedingber — APPROVED** [review]

**Pull request merged**